src/app/Library/CrudPanel/Traits/Validation.php

```diff
@@ -306,16 +180,25 @@ public function setRequiredFields($classOrRulesArray)
         }
 
         if (count($rules)) {
-            foreach ($rules as $key => $rule) {
-                if (
-                    (is_string($rule) && strpos($rule, 'required') !== false && strpos($rule, 'required_') === false) ||
-                    (is_array($rule) && array_search('required', $rule) !== false && array_search('required_', $rule) === false)
-                ) {
-                    if (Str::contains($key, '.')) {
-                        $key = Str::dotsToSquareBrackets($key, ['*']);
-                    }
+            foreach ($rules as $key => $validationRules) {
+                if (is_string($validationRules)) {
+                    $validationRules = explode('|', $validationRules);
+                }
+                foreach ($validationRules as $rule) {
+                    if (is_a($rule, ValidBackpackUpload::class, true)) {
```

Comment:
From what I can tell... this is tying a core function (`public function setRequiredFields()`) to a custom validator we've just created - `ValidBackpackUpload` 👀

![red-flag-stop-gif-by-richard-childress-racing](https://user-images.githubusercontent.com/1032474/233598758-d017adb1-a953-4269-898c-d83ec0c8875f.gif)

Could you please explain why this is needed @pxpm ? I guess it's related to what you tried to explain to me on Discord, but I didn't understand there either 😀

1) ELI5 why this is needed please. 
2) Why do we consider `ValidBackpackUpload` is the same thing as `required` - to show the asterisk? 
3) Why don't people have to do both `required` and `ValidBackpackUpload` when that is the case? Why does `ValidBackpackUpload` also mean required? What about when you want it to be a valid upload, but only if present? Like if you would combine `nullable` and `ValidBackpackUpload` - in my mind?

Reply:
Hey @tabacitu 

`ValidBackpackUpload` is the "validation rule" that developers define. There "real rules" are inside `->arrayRules()` when the rule is a `ValidBackpackUpload` rule (both `ValidUploadMultiple` and `ValidDropzone` are extensions of the `ValidBackpackUpload` rule).

So what I am doing here, is checking if the rule is a `ValidBackpackUpload`, in case it is, we check the "requiredeness" 😎 inside the rules in `arrayRules()`. 

```php
'attachments' => ['required', ValidUploadMultiple::make()->arrayRules('min:2|max:3')->fileRules(...)]
'attachments' => [ValidUploadMultiple::make()->arrayRules('required|min:2|max:3')->fileRules(...)] // the same as above

// answering your question: I want it to be a valid upload, but only if present.
'attachments' => [ValidUploadMultiple::make()->fileRules(...)],
//or when present min 1 file
'attachments' => [ValidUploadMultiple::make()->arrayRules('min:1|max:3')->fileRules(...)],
'attachments' => ['min:1', 'max:3', ValidUploadMultiple::make()->fileRules(...)], // the same as above
``` 

In that foreach we are iterating the `attachments` validation rules, so for the example: 
`'attachments' => ['required', ValidUploadMultiple::make()->arrayRules('min:2|max:3')->fileRules(...)]`

- the first iteration would rule `required` rule and follow the previous check.
- the second and last would be `ValidUploadMultiple` rule, in that case we will check if the `required` is inside the `arrayRules`

This does not automatically set required because it's a `ValidBackpackUpload`. It handles them and set required if they are required from the array rules  👍

Reply:
I am not sure I've followed you here. 

Are you suggesting something like renaming `ValidBackpackUpload`  to `ValidArrayData` ? So that we can have:
```php
class ValidUploadMultiple extends ValidArrayData {}
class ValidRepeatable extends ValidArrayData {}
```
If that's the case I agree with you I guess. But ValidArrayData should not be the "rule", only the `ValidUploadMultiple` etc are rules correct?
Like:
```php
class ValidArrayData { }
class ValidUploadMultiple extends ValidArrayData implements ValidationRule, DataAwareRule, ValidatorAwareRule {}
```
Cheers

Reply:
I think if we do this the `ValidUploadMultiple` class would need to have `getData()`, `getValidator()` and the "custom validator methods" that are used in other validators like `ValidDropzone` and that would bloat up the files so much. 

So I think we should still keep the `ValidArrayData` as the "validation rule", and the others extend it. 

```php
class ValidArrayData implements ValidationRule, DataAwareRule, ValidatorAwareRule {}
class ValidUploadMultiple extends ValidArrayData 
```

Reply:
Yup - seems like we're on the same page. The only bad feeling I have is about the name - maybe we call it directly `ValidArray` instead of `ValidArrayData`? "data" doesn't really say anything, does it?